snapshots/manager.go

```diff
@@ -0,0 +1,259 @@
+	err      error // if non-nil, restore errored
+}
+
+// Manager manages snapshot and restore operations for an app, making sure only a single
+// long-running operation is in progress at any given time, and provides convenience methods
+// mirroring the ABCI interface.
+//
+// Although the ABCI interface (and this manager) passes chunks as byte slices, the internal
+// snapshot/restore APIs use IO streams (i.e. chan io.ReadCloser), for two reasons:
+//
+// 1) In the future, ABCI should support streaming. Consider e.g. InitChain during chain
+//    upgrades, which currently passes the entire chain state as an in-memory byte slice.
+//    https://github.com/tendermint/tendermint/issues/5184
+//
+// 2) io.ReadCloser streams automatically propagate IO errors, and can pass arbitrary
+//    errors via io.Pipe.CloseWithError().
+type Manager struct {
+	store  *Store
+	target types.Snapshotter
+
+	mtx                sync.Mutex
+	operation          operation
+	chRestore          chan<- io.ReadCloser
+	chRestoreDone      <-chan restoreDone
+	restoreChunkHashes [][]byte
+	restoreChunkIndex  uint32
+}
+
+// NewManager creates a new manager.
+func NewManager(store *Store, target types.Snapshotter) *Manager {
+	return &Manager{
+		store:  store,
+		target: target,
+	}
+}
+
+// begin starts an operation, or errors if one is in progress. It manages the mutex itself.
+func (m *Manager) begin(op operation) error {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	return m.beginLocked(op)
+}
+
+// beginLocked begins an operation while already holding the mutex.
+func (m *Manager) beginLocked(op operation) error {
+	if op == opNone {
+		return sdkerrors.Wrap(sdkerrors.ErrLogic, "can't begin a none operation")
+	}
+	if m.operation != opNone {
+		return sdkerrors.Wrapf(sdkerrors.ErrConflict, "a %v operation is in progress", m.operation)
+	}
+	m.operation = op
+	return nil
+}
+
+// end ends the current operation.
+func (m *Manager) end() {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	m.endLocked()
+}
+
+// endLocked ends the current operation while already holding the mutex.
+func (m *Manager) endLocked() {
+	m.operation = opNone
+	if m.chRestore != nil {
+		close(m.chRestore)
+		m.chRestore = nil
+	}
+	m.chRestoreDone = nil
+	m.restoreChunkHashes = nil
+	m.restoreChunkIndex = 0
+}
+
+// Create creates a snapshot and returns its metadata.
+func (m *Manager) Create(height uint64) (*types.Snapshot, error) {
+	if m == nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrLogic, "no snapshot store configured")
+	}
+	err := m.begin(opSnapshot)
+	if err != nil {
+		return nil, err
+	}
+	defer m.end()
+
+	latest, err := m.store.GetLatest()
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "failed to examine latest snapshot")
+	}
+	if latest != nil && latest.Height >= height {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrConflict,
+			"a more recent snapshot already exists at height %v", latest.Height)
+	}
+
+	chunks, err := m.target.Snapshot(height, types.CurrentFormat)
```

Comment:
When we change the snapshot format, we could conceivably want nodes to be able to generate snapshots both in the old and new format, to support both old and new state sync clients. Currently this is entirely academic, since we only have a single format - can hardcode it if you'd prefer.

Reply:
This is fine for now then.